In Docker files modify Debian image with latest stable Bookworm release

hammera opened the pull request:
Hi Boys,

In June 11 Debian 12.0 Bookworm are released, this is the new stable release until five year support (three year security support and two year LTS support).
I updated the three docker files the Debian:latest image referrence with Debian 12.0 point release (of course I pinned the sha256sum value) the correct value.

In Dockerfile, Dockerfile.win32 files enough to change the image reference value, but the Dockerfile.win64 need me change the wine64 related command line parameter related lines.

In Debian Bookworm release wine64 command awailable into the /usr/lib/wine/wine64 path, not available with /usr/bin.
Please carefully review this change, and if you would like, feel free merge this change to the master branch.

My machine ran make distwin command succesfull, both 32 bit Windows version and 64 bit Windows version launched in Wine correct.

A question:
In Liblouis UTDML docker files need replace the image names?
Into the Liblouis UTDML Github repository Dockerfile.win32 and Dockerfile.win64 uses debian:latest image name, with now pulls with Debian 12.0 stable release based image.
In Dockerfile the image name defined with following from line with Liblouis UTDML Github repository:
FROM liblouis/liblouis

If you would like, I will doing a pull request in Liblouis UTDML repo the image replacement related if you would like same pinned style images, based in Liblouis repository Dockerfile.win32 and Dockerfile.win64 style image names.

Attila

hammera commented:
Nice, all checks are passed.

Attila

hammera commented:
I fixed sha256sum value the normal Dockerfile (not Windows builds related docker files).
I simple forgot paste new sha256sum value, so this is a cosmetic fix, not affects Windows builds.

Attila

egli commented:
> In Liblouis UTDML docker files need replace the image names? Into the Liblouis UTDML Github repository Dockerfile.win32 and Dockerfile.win64 uses debian:latest image name, with now pulls with Debian 12.0 stable release based image. 

That is probably OK, they will just use the newest image.  The change you did to the path of the wine binary might be needed.

> In Dockerfile the image name defined with following from line with Liblouis UTDML Github repository: FROM liblouis/liblouis

That should be OK. It just uses an image that already has a build of liblouis in it.
 
> If you would like, I will doing a pull request in Liblouis UTDML repo the image replacement related if you would like same pinned style images, based in Liblouis repository Dockerfile.win32 and Dockerfile.win64 style image names.

A PR for utdml would be nice, but I don't think we need the pinning there. It does add some security but I think the benefit doesn't outweigh the added complexity (at least for utdml). All we need in the utdml project is to have working docker images so we can cross-compile utdml.

hammera commented:
Hi Chris,

Thanks the answer. I will doing the proper pull request for Liblouis UTDML related.
What do you think? What we doing if Debian future publishing for example 12.1, 12.2, 12.3 point releases?
We alwais changing the sha256sum values in Liblouis docker files after end of debian:latest line?
If we keep the now used sha256sum value, we will be pinned future with 12.0 based image.

If remove the sha256sum value, we pinned always the latest Debian stable release (possible only two or three years later will be come on the next stable main release), but we always pinned the latest stable point release.

Attila

hammera commented:
Chris, I have got a good news:
Now I tested with make distwin command in Liblouis UTDML related.
Both two builds happened succesfully (independent with Debian:latest docker image based now with Debian 12, not Debian 11).
So, main debian version change doesn't affects Liblouis Utdml 32 bit and 64 bit windows build quality level.
Not need doing Wine related any pull request for Liblouis Utdml related, because in Docker files only wine64 package is installed, but not execute the Docker file with wine64 command.

Attila

hammera commented:
Chris, Because an another merged  pull request Debian image in Docker files are updated with latest sha256sum value, I rebased the branch and force-pushed the branch content.
This change results keeped only the Dockerfile.wine64 file the wine64 binary path related change.
I finalized this PR, pull request is safe merge with master branch if you have a little time.

Attila

egli commented:
The win32 and win64 Dockerfiles are quite important as we use them to provide cross-compiled images.

Now it seems that we mistakenly upgraded to a newer version of Debian in #1413. This means that we can no longer provide cross-compile binaries of liblouis.

So I tried to revert the Dockerfiles to a previous version of Debian (bullseye). It all compiles fine but the tests for both w32 and w64 fail. Did they always fail? This would indicate that the pre-built binaries do not even work.

Here's the output of me running `docker build -t liblouis/liblouis-win64 -f Dockerfile.win64 .`
```
Step 21/22 : ADD tests/yaml/letterDefTest_harness.yaml .
 ---> ade0d5878e1c
Step 22/22 : RUN wine64 bin/lou_checkyaml.exe letterDefTest_harness.yaml
 ---> Running in ea92a837e931
wine: created the configuration directory '/root/.wine'
0012:err:ole:marshal_object couldn't get IPSFactory buffer for interface {00000131-0000-0000-c000-000000000046}
0012:err:ole:marshal_object couldn't get IPSFactory buffer for interface {6d5140c1-7436-11ce-8034-00aa006009fa}
0012:err:ole:StdMarshalImpl_MarshalInterface Failed to create ifstub, hres=0x80004002
0012:err:ole:CoMarshalInterface Failed to marshal the interface {6d5140c1-7436-11ce-8034-00aa006009fa}, 80004002
0012:err:ole:get_local_server_stream Failed: 80004002
000b:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
000b:err:winediag:nodrv_CreateWindow Make sure that your X server is running and that $DISPLAY is set correctly.
0014:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
0014:err:winediag:nodrv_CreateWindow Make sure that your X server is running and that $DISPLAY is set correctly.
0014:err:ole:apartment_createwindowifneeded CreateWindow failed with error 0
0014:err:ole:apartment_createwindowifneeded CreateWindow failed with error 0
0014:err:ole:marshal_object couldn't get IPSFactory buffer for interface {00000131-0000-0000-c000-000000000046}
0014:err:ole:apartment_createwindowifneeded CreateWindow failed with error 14007
0014:err:ole:StdMarshalImpl_MarshalInterface Failed to create ifstub, hres=0x800736b7
0014:err:ole:CoMarshalInterface Failed to marshal the interface {6d5140c1-7436-11ce-8034-00aa006009fa}, 800736b7
0014:err:ole:get_local_server_stream Failed: 800736b7
0010:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
0010:err:winediag:nodrv_CreateWindow Make sure that your X server is running and that $DISPLAY is set correctly.
Could not find Wine Gecko. HTML rendering will be disabled.
it looks like wine32 is missing, you should install it.
multiarch needs to be enabled first.  as root, please
execute "dpkg --add-architecture i386 && apt-get update &&
apt-get install wine32"
0025:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
0025:err:winediag:nodrv_CreateWindow Make sure that your X server is running and that $DISPLAY is set correctly.
Could not find Wine Gecko. HTML rendering will be disabled.
wine: configuration in L"/root/.wine" has been updated.
source unknown: error: Character can not be (indirectly) based on itself.
source unknown: error: Character can not be (indirectly) based on itself.
Translation failed.
SUCCESS (8 tests, 0 failures)
Removing intermediate container ea92a837e931
 ---> ccf20fbb8475
Successfully built ccf20fbb8475
Successfully tagged liblouis/liblouis-win64:latest
```

Oddly enough the test suite thinks it passed, but the `source unknown: error: Character can not be (indirectly) based on itself.` is definitely an error coming from liblouis.

When running for w32 I get the same problem: `docker build -t liblouis/liblouis-win32 -f Dockerfile.win32 .`

```
Step 22/23 : ADD tests/yaml/letterDefTest_harness.yaml .
 ---> d08eddb47906
Step 23/23 : RUN wine bin/lou_checkyaml.exe letterDefTest_harness.yaml
 ---> Running in 1413861a6de0
wine: created the configuration directory '/root/.wine'
0050:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
0050:err:winediag:nodrv_CreateWindow L"The explorer process failed to start."
0050:err:systray:initialize_systray Could not create tray window
0050:err:ole:StdMarshalImpl_MarshalInterface Failed to create ifstub, hr 0x80004002
0050:err:ole:CoMarshalInterface Failed to marshal the interface {6d5140c1-7436-11ce-8034-00aa006009fa}, hr 0x80004002
0050:err:ole:apartment_get_local_server_stream Failed: 0x80004002
0040:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
0040:err:winediag:nodrv_CreateWindow L"Make sure that your X server is running and that $DISPLAY is set correctly."
0048:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
0048:err:winediag:nodrv_CreateWindow L"Make sure that your X server is running and that $DISPLAY is set correctly."
0048:err:ole:apartment_createwindowifneeded CreateWindow failed with error 3
0048:err:ole:apartment_createwindowifneeded CreateWindow failed with error 0
0048:err:ole:apartment_createwindowifneeded CreateWindow failed with error 14007
0048:err:ole:StdMarshalImpl_MarshalInterface Failed to create ifstub, hr 0x800736b7
0048:err:ole:CoMarshalInterface Failed to marshal the interface {6d5140c1-7436-11ce-8034-00aa006009fa}, hr 0x800736b7
0048:err:ole:apartment_get_local_server_stream Failed: 0x800736b7
0048:err:ole:start_rpcss Failed to open RpcSs service
002c:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
002c:err:winediag:nodrv_CreateWindow L"Make sure that your X server is running and that $DISPLAY is set correctly."
008c:err:winediag:nodrv_CreateWindow Application tried to create a window, but no driver could be loaded.
008c:err:winediag:nodrv_CreateWindow L"Make sure that your X server is running and that $DISPLAY is set correctly."
wine: configuration in L"/root/.wine" has been updated.
source unknown: error: Character can not be (indirectly) based on itself.
source unknown: error: Character can not be (indirectly) based on itself.
Translation failed.
SUCCESS (8 tests, 0 failures)
Removing intermediate container 1413861a6de0
 ---> 8cc0bbd95375
Successfully built 8cc0bbd95375
Successfully tagged liblouis/liblouis-win32:latest
```
I'm a bit stumped as to what I should do now.

hammera commented:
In june 5 I think happened following commit, commit id is following:
56f26d8b39b0d9370e84a5f176c3c5af88d4182b
The upgrade is need because Debian Jessie (with uses previous Liblouis) already not available.
Debian:latest named image pulls always with latest stable image, if you would like for example Debian Bullseye based version, you need use the pull line with debian:bullseye style line (and need perhaps remove the sha256sum value).
I think wine requires usual with running x server, but I am not full sure this.
Oldest time after june 5 I think real environment we Tried Norbert the cross-compiled Liblouis version, worked fine.
If you would like, next week will be I generating latest image a cross-compiled build, and ran a machine with have X server and wine.

Attila

hammera commented:
Wait Chris a little bit, I see now an interesting thing:
When I looked Linux with tools/lou_checkyaml tests/yaml/letterdefs_harness.yaml command, I see following errors:
```
source unknown: error: Character can not be (indirectly) based on itself.
source unknown: error: Character can not be (indirectly) based on itself.
Translation failed.
SUCCESS (8 tests, 0 failures)
```
Why using we this buggy this tests both Linux and Windows systems, or what means this mistific errors before the success results?
Not perhaps better we use an another working tests (replace the letterdefs_harness.yaml test in Dockerfile.win32 and Dockerfile.win64) for example the tests/braille-specs/spaces.yaml) to test cross-compiled Liblouis cross-compiled build works or not?

With not reverted upgraded Debian Bookworm version the 32 bit Windows version compiled right, and the 64 bit version are not compiled.
I redirected the entire compilation output. 64 bit version compilation ends with the 22 step with following error, because wine path are not changed with the Dockerfile.win64 command:
```
Step 22/22 : RUN wine64 bin/lou_checkyaml.exe letterDefTest_harness.yaml
 ---> Running in d019675a1d0d
[91m/bin/sh: 1: wine64: not found
```
This is true because wine64 lives with /usr/lib/wine directory (I added this modification the Dockerfile.wine64 this PR but final not merged this modification).

I think other Wine related diagnostic errors are not related with Liblouis cross-compiled build, but related the X Server configuration. Docker environment default not have running X server, but yaml testing in wine absolute not need a running full capacity X server to run lou_checkyaml.exe command.
I reading now following documentation to setup a full capacity Docker wine environment:
https://leimao.github.io/blog/Docker-Wine/

What your openion?
I trying replace the docker files with letterdefs_harness.yaml to spaces.yaml? What debian version you would like final during cross-compilation? Debian Bookworm or Debian Bullseye?

Attila

hammera commented:
Chris, I little redesigned the Dockerfile.win32 and Dockerfile.win64, and replaced letterdefs_harness.yaml the spaces.yaml file to we ensure works right the cross-compiled 32 bit and 64 bit Liblouis Windows versions.
I have got good news.
Look following diff output, this variant is debian:latest based images compatible, and generated succesfully cross-compiled 32 bit and 64 bit Windows builds my feature branch. This branch is based with Debian Bookworm stable release with have support until 2028:
```
diff --git a/Dockerfile.win32 b/Dockerfile.win32
index d306f6bb..334a4aa4 100644
--- a/Dockerfile.win32
+++ b/Dockerfile.win32
@@ -73,10 +73,11 @@ RUN apt-get update && dpkg --add-architecture i386 && apt-get update && apt-get
    && rm -rf /var/lib/apt/lists/*
 
 ENV SRCDIR=/usr/src/liblouis
+env LOUIS_TABLEPATH=/usr/src/liblouis/share/liblouis/tables
 WORKDIR ${SRCDIR}
 COPY --from=builder ${SRCDIR}/liblouis.zip .
 RUN unzip liblouis.zip
 
 # just run any old self-contained yaml test that doesn't need any env setup
-ADD tests/yaml/letterDefTest_harness.yaml .
-RUN wine bin/lou_checkyaml.exe letterDefTest_harness.yaml
+ADD tests/braille-specs/spaces.yaml .
+RUN wine bin/lou_checkyaml.exe spaces.yaml
diff --git a/Dockerfile.win64 b/Dockerfile.win64
index 620f7b66..1fb68cd0 100644
--- a/Dockerfile.win64
+++ b/Dockerfile.win64
@@ -19,7 +19,7 @@ RUN apt-get update && apt-get install -y \
 
 ARG LIBYAML_VERSION=0.1.4
 ENV HOST=x86_64-w64-mingw32 PREFIX=/usr/build/win64 SRCDIR=/usr/src/
-
+env LOUIS_TABLEPATH=/usr/src/liblouis/share/liblouis/tables
 # Build and install libyaml
 WORKDIR ${SRCDIR}
 RUN curl -L https://github.com/yaml/libyaml/archive/${LIBYAML_VERSION}.tar.gz | tar zx
@@ -60,10 +60,11 @@ RUN apt-get update && apt-get install -y \
    && rm -rf /var/lib/apt/lists/*
 
 ENV SRCDIR=/usr/src/liblouis
+env LOUIS_TABLEPATH=/usr/src/liblouis/share/liblouis/tables
 WORKDIR ${SRCDIR}
 COPY --from=builder ${SRCDIR}/liblouis.zip .
 RUN unzip liblouis.zip
 
 # just run any old self-contained yaml test that doesn't need any env setup
-ADD tests/yaml/letterDefTest_harness.yaml .
-RUN /usr/lib/wine/wine64 bin/lou_checkyaml.exe letterDefTest_harness.yaml
+ADD tests/braille-specs/spaces.yaml .
+RUN /usr/lib/wine/wine64 bin/lou_checkyaml.exe spaces.yaml
```

I replaced the tests/yaml/letterDefTest_harness.yaml file, because both Linux and Windows generated you quoted error messages. I choosed tests/braille-specs/spaces.yaml file, because have possibility verify this file we get equals good test results both Linux and Windows build.
In Linux when I test spaces.yaml, all 1629 tests passed right the cross-compiled builds too.
This is a good measurement way to we always ensure works right the cross-compiled build or not.

The X server related error messages is not interesting I think with Liblouis related, lou_checkyaml.exe is a command-line program and doesn't need any X server to ran this utility.
Looks following output parts what happening if I running make distwin command and building both 32 and 64 bit Liblouis Windows versions, and how many test results passed if I testing spaces.yaml file:
```
[0m[91mTables have not been indexed yet. Indexing LOUIS_TABLEPATH.
[0m[91mBest match: /usr/src/liblouis/share/liblouis/tables/ga-g1.utb (15)
[0m[91mBest match: /usr/src/liblouis/share/liblouis/tables/ga-g2.ctb (15)
[0m[91mDuplicate emphasis class: italic
Duplicate emphasis class: underline
Duplicate emphasis class: bold
[0m[91m3 warnings issued
[0m[91mDuplicate emphasis class: italic
Duplicate emphasis class: underline
Duplicate emphasis class: bold
[0m[91m3 warnings issued
[0m[91mDuplicate emphasis class: italic
Duplicate emphasis class: underline
Duplicate emphasis class: bold
Duplicate emphasis class: script
Duplicate emphasis class: transnote
Duplicate emphasis class: trans1
Duplicate emphasis class: trans2
Duplicate emphasis class: trans3
Duplicate emphasis class: trans4
Duplicate emphasis class: trans5
[0m[91m10 warnings issued
[0mSUCCESS (1629 tests, 0 failures)
Removing intermediate container 3672c5867406
 ---> 0a37dff578f6
Successfully built 0a37dff578f6
Successfully tagged liblouis/win32:latest
```

64 bit build part output result:
```
[0m[91mTables have not been indexed yet. Indexing LOUIS_TABLEPATH.
[0m[91mBest match: /usr/src/liblouis/share/liblouis/tables/ga-g1.utb (15)
[0m[91mBest match: /usr/src/liblouis/share/liblouis/tables/ga-g2.ctb (15)
[0m[91mDuplicate emphasis class: italic
Duplicate emphasis class: underline
Duplicate emphasis class: bold
[0m[91m3 warnings issued
[0m[91mDuplicate emphasis class: italic
Duplicate emphasis class: underline
Duplicate emphasis class: bold
[0m[91m3 warnings issued
[0m[91mDuplicate emphasis class: italic
Duplicate emphasis class: underline
Duplicate emphasis class: bold
Duplicate emphasis class: script
Duplicate emphasis class: transnote
Duplicate emphasis class: trans1
Duplicate emphasis class: trans2
Duplicate emphasis class: trans3
Duplicate emphasis class: trans4
Duplicate emphasis class: trans5
[0m[91m10 warnings issued
[0mSUCCESS (1629 tests, 0 failures)
Removing intermediate container 7c7fc1794129
 ---> 21499a4082c9
Successfully built 21499a4082c9
Successfully tagged liblouis/win64:latest
```

The unwanted escape sequences before command outputs is happening because i put entire compilation output a file the make distwin >anything.txt command.

So, with Debian Bookworm stable release we have possibility to generate cross-compiled good working Liblouis versions, this is absolute showed and proofed the spaces.yaml test result (all 1629 space tests passed).

What we doing? I commit this diff patch the feature branch, or we migrate back with Debian Bullseye and keep the spaces.yaml related modification?

Spaces.yaml is a very good measurement to we always ensure both Linux and Windows compatible cross-compiled builds works right.

Attila

hammera commented:
Chris, Again changed with debian:latest sha256sum value?
In Dockerfile aug 11 you replaced sha256sum value right from with debian:latest@sha256:d568e251e460295a8743e9d5ef7de673c5a8f9027db11f4e666e96fb5bed708e to debian:latest@sha256:f2150eba68619015058b26d50e47f9fba81213d1cb81633be7928c830f72d180, but this replacement not happened with Dockerfile.win32 and Dockerfile.win64.
If very often changing the sha256sum value, we need pinning the sha256sum value a specific maintenance version point, for example this now you used sha256sum value? If yes, this situation before we install dependencies perhaps need always doing an apt-get update and an apt-get full-upgrade command, to always happens the stable release update before cross-compiled builds generation.

I update the sha256sum values my feature branch the Dockerfile.win32 and Dockerfile.win64 files, or you will be assigning this little one line update, similar with normal Linux compatible Dockerfile?

Attila

egli commented:
I tried to reproduce your results but it fails for me. I even tried to go back to jessie to make it work again. With buster I get the same `source unknown: error: Character can not be (indirectly) based on itself.` error message

Unfortunately I do not have any more time to work on this at the moment

hammera commented:
tOk, thanks Chris.
What can I help for you this issue related?
Spaces.yaml test file migration is not a good option (me showed diff output part)?
Have a problem perhaps letterDefTest_harness.yaml file possible, because in Normal Ubuntu system I get same result without compilation when I use tools/lou_checkyaml  tests/yaml/letterDefTest_harness.yaml command:
```
source unknown: error: Character can not be (indirectly) based on itself.
source unknown: error: Character can not be (indirectly) based on itself.
Translation failed.
SUCCESS (8 tests, 0 failures)
```

The debian:jessie image are impossible to download any packages, because when you trying doing an apt-get update command, repository list are already not accessible (produce 404 errors).
debian:buster, debian:bullseye, debian:bookworm (or debian:latest, this is now equals because the latest tag shows always the actual stable release), debian:testing, or debian:sid is a direction step option, but I not suggest debian:testing or debian:sid if you not would like lot of maintenance and unwanted problems if happens an automatic package component upgrade.
If we go back debian:bullseye (the oldstable repository), we have support period until 2026 april or may before bullseyee repositoryes beginning drops 404 errors. Possible more time, because jessie begin dropping 404 errors I think june or jul this year. :-):-)
So, if you have possibility look short time with I quoted diff part the spaces.yaml related if want replace test part the letterDefTest_harness.yaml part, I playing the docker files if tell me what the target build system with we optimal situation wants use final with cross-compile (Buster, Bullseye or Bookworm latest stable release).
When I doed the saturday doed test the modified but not committed Dockerfile.win32 and Dockerfile.win64 files, when I ran make distwin command, I right get 32 bit and 64 bit Windows builds zipped versions.
For example, when I unzipped the versions, in wine launch right lou_translate.exe command both 32 bit and 64 bit versions, of course I need define LOUIS_TABLEPATH variable with proper value my home directory with I unzipped the proper architecture Windows version (depending 32 bit or 64 bit).

Attila

hammera commented:
Chris, well, I little hammered the Dockerfile.win32, Dockerfile.win64 files, but before commit and push I need ran a docker build -f Dockerfile.win64 . command.
When I analyzed the 32 bit build log, I see following result, so all required tests with need ran the Windows version passed right before generating Liblouis 32 bit version.
Look following log result part from 32 bit Windows version local Docker build, based with you august 11 committed debian:latest sha256sum value version:
```
Testsuite summary for Liblouis 3.26.0
============================================================================
# TOTAL: 199
# PASS:  199
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
```
In Windows cross-compilation real good the 199 total tests number?
In my Linux box when I ran the make check command the test directory, I see following results:
```
===========================================
   Liblouis 3.26.0: tests/test-suite.log
===========================================

# TOTAL: 203
# PASS:  202
# SKIP:  0
# XFAIL: 1
# FAIL:  0
# XPASS: 0
# ERROR: 0
```

So, remaining me following:
1. test manual with Wine the 32 bit version, for example if I ran bin/lou_translate.exe command in wine the hu-hu-g1.ctb table when setted the LOUIS_TABLEPATH enwironment value with right way, ran my real Linux box with bin/lou_checkyaml.exe for example the large hungarian dictionary hu-hu-g1_dictionary_special_consonants.yaml.
2. Do a 64 build verification.
3. Test same with 64 bit generated build with equals mode with I described the first step.

Attila

hammera commented:
Very fine, 32 bit local Windows build passed the manual test.
I unzipped the generated liblouis-3.26.0-win32.zip with my home directory, and exported LOUIS_TABLEPATH environment value with following command:
```
export LOUIS_TABLEPATH=z:\\home\\hammera\\liblouis\\share\\liblouis\\tables
```
After this, I copyed from Liblouis source tree with tests/braille-specs/hu-hu-g1_dictionary_special_consonants.yaml file, and tests/braille-specs/hu-hu-g1_harness.yaml test files the liblouis directory.
When I ran wine bin/lou_checkyaml.exe hu-hu-g1_harness.yaml command, I see following result my terminal, this is the right result with Wine compatible build:
```
Tables have not been indexed yet. Indexing LOUIS_TABLEPATH.
Best match: z:\home\hammera\liblouis\share\liblouis\tables/hu.tbl (16)
SUCCESS (1275 tests, 0 failures)
```
And this is the large dictionary test result:
```
Tables have not been indexed yet. Indexing LOUIS_TABLEPATH.
Best match: z:\home\hammera\liblouis\share\liblouis\tables/hu.tbl (16)
SUCCESS (85932 tests, 0 failures)
```
I tried a wine bin/lou_translate.exe hu-hu-g1.ctb command, and type some text. Translation happened right.
So, 32 bit builded version is good.

Now running the 64 bit Windows build generation my system.

Attila

hammera commented:
This is the 64 bit Windows version manual test result, test steps is same:
I unzipped the 64 bit version my home directory the liblouis folder. Because the LOUIS_TABLEPATH environment value is still exported with Wine compatible way, I simple ran wine bin/lou_translate.exe --version command.
This is the output result:
```
wine bin/lou_translate.exe --version
Z:\home\hammera\liblouis\bin\lou_translate.exe (Liblouis) 3.26.0
Copyright (C) 2022 ViewPlus Technologies, Inc. and JJB Software, Inc.
License GPLv3+: GNU GPL version 3 or later <https://gnu.org/licenses/gpl.html>.
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.

Written by John J. Boyer.
```
After copyed to /home/hammera/liblouis the previous comment used test files, I repeated other tests. commands. I ran wine bin/lou_checkyaml.exe hu-hu-g1_harness.yaml and wine bin/lou_checkyaml.exe hu-hu-g1_dictionary_special_consonants.yaml commands, both two tests I see the previous comment wrote test results.

I think ready the Windows build part related sincronization.

Attila

hammera commented:
Chris, I doed two commit final this day.
The first committed change is syncing only with Dockerfile.win32 and Dockerfile.win64 with you aug 11 committed sha256sum value.
The first this day committed change is a small diff, easying to review.

The second commit does the magic thing.
I added following commit description:
```
Replace tests/yaml/letterDefTest_harness.yaml with tests/braille-specs/spaces.yaml with Dockerfile.win32 and Dockerfile.win64 files
The tests/yaml/letterDefTest_harness.yaml produce following errors before possible false positive success result the build log both Linux and Windows builds:
source unknown: error: Character can not be (indirectly) based on itself.
source unknown: error: Character can not be (indirectly) based on itself.
Translation failed.
SUCCESS (8 tests, 0 failures)
I replaced the tests/yaml/letterDefTest_harness.yaml with tests/braille-specs/spaces.yaml, because if the Dockerfile.win32 and Dockerfile.win64 the proper LOUIS_TABLEPATH environment value is defined, the test executes succesfull during cross-compilation when executing wine bin/lou_checkyaml.exe spaces.yaml command, bot 32 bit and 64 Wine versions.
All the 1629 space tests passed, similar in Linux build.
This is a very good measurement to we always ensure works right the cross-compiled 32 bit and 64 bit Windows builds.
```

This commit I fixed only 64 bit Windows build related Dockerfile.win64 file the wine 64 path.
So, the second commit fix is small similar, easy review.
The human tests are confirmed this way works right the 32 bit and 64 bit Windows version build (see the previous comments related test results), you already fixed Linux compatible Dockerfile.
So, we I think ready this task related.

Look the small diff with me fixes the Windows builds related task both two architectures, the diff output is 38 lines only:
```
diff --git a/Dockerfile.win32 b/Dockerfile.win32
index 3f1ba99e..53ca19d1 100644
--- a/Dockerfile.win32
+++ b/Dockerfile.win32
@@ -73,10 +73,11 @@ RUN apt-get update && dpkg --add-architecture i386 && apt-get update && apt-get
    && rm -rf /var/lib/apt/lists/*
 
 ENV SRCDIR=/usr/src/liblouis
+env LOUIS_TABLEPATH=/usr/src/liblouis/share/liblouis/tables
 WORKDIR ${SRCDIR}
 COPY --from=builder ${SRCDIR}/liblouis.zip .
 RUN unzip liblouis.zip
 
-# just run any old self-contained yaml test that doesn't need any env setup
-ADD tests/yaml/letterDefTest_harness.yaml .
-RUN wine bin/lou_checkyaml.exe letterDefTest_harness.yaml
+# run spaces.yaml test to we ensure cross-compile build works right with 32 bit Windows versions
+ADD tests/braille-specs/spaces.yaml .
+RUN wine bin/lou_checkyaml.exe spaces.yaml
diff --git a/Dockerfile.win64 b/Dockerfile.win64
index 7c5589df..ed1a17c1 100644
--- a/Dockerfile.win64
+++ b/Dockerfile.win64
@@ -60,10 +60,11 @@ RUN apt-get update && apt-get install -y \
    && rm -rf /var/lib/apt/lists/*
 
 ENV SRCDIR=/usr/src/liblouis
+env LOUIS_TABLEPATH=/usr/src/liblouis/share/liblouis/tables
 WORKDIR ${SRCDIR}
 COPY --from=builder ${SRCDIR}/liblouis.zip .
 RUN unzip liblouis.zip
 
-# just run any old self-contained yaml test that doesn't need any env setup
-ADD tests/yaml/letterDefTest_harness.yaml .
-RUN /usr/lib/wine/wine64 bin/lou_checkyaml.exe letterDefTest_harness.yaml
+# run spaces.yaml test to we ensure cross-compile build works right with 64 bit Windows versions
+ADD tests/braille-specs/spaces.yaml .
+RUN /usr/lib/wine/wine64 bin/lou_checkyaml.exe spaces.yaml
```

Attila

hammera commented:
Boys, I doed some optimizations with Dockerfile, Dockerfile.win32, Dockerfile.win64 files.
After latest debian image pull usual previous we doed only apt-get update, and not doed
apt-get -y dist-upgrade command before components installation (build-depends installation with Liblouis compilation related).
This is not good, because between maintenance Debian points releases have more
updated components with contains security fixes or other perhaps important updates with affects only the basic debian:latest pulled system image.
I corrected this with the three Docker files, and after larger components
installation I always running apt-get clean command to decreasing build image sizes.
This way optimization the normal Linux compatible generated build image is more smaller (about only 700 MB), Win32 and Win64 Docker builded image is estimated 3,2 GB/build.

This commit is review ready, I will not doing any commit this PR related until september 12 (the 3.27 Liblouis version release date), except not have review requested changes.
Chris, @bertfrees, I wish you good vacation.

hammera commented:
I forgot:
All checks passedright. I tested local in Docker the Linux compatible building command and Dockerfile.win32 generated cross-compile build, tomorrow morning I have possibility to test a local 64 bit Windows build generation test.

hammera commented:
Boys, what will be the cross-compile Docker images final direction?
Chris I think since three weeks ago already opened a Bullseye pinned PR with #1422 with have an approve from Bert, if I remember right an automatic dependabot user opened a pull request with update latest committed Docker images with actual sha256sum values, pr number is #1441 (this PR require a review, not happened yet a review).
And final, me have a working solution and tested cross-compile build the latest debian based images too, but now I little confused when I seed this pull requests the PR list. :-):-)
So, please tell me what the final image (bookworm or bullseye version) with I need actualizing my feature branch, because my branch need doing a rebase operation (the branch is already outdated, not the Docker image files, Docker files is right).
Note, only a question:
Realy we need applying the sha256sum number in the Docker files? Not better if we uses debian:latest or debian:bullseye image names simple when pulls the proper base system, similar with Liblouis UTDML cross-compile docker files?
This method is more simple, and after we final committing the proper Docker files, more years not need doing any change the Docker files related.
Until september 12 we have more times to finalize Docker files, I wlcome assigning this change work for me if you tell the proper direction and base system distribution version (bullseye or Debian Bookworm latest stable version.
Monday I am not in home possible and not near with internet, but tuesday I have possibility to working with Docker files and test the cross-compilation before merge.

Excuse me if my english is not always right, hopefully everithing is understandable,

Attila

egli commented:
We're dropping this from the release for now. We'll have to do some more investigation:

1. Is the problem just with running the tests with wine? Maybe the cross-compiled exes are totally fine?
2. Maybe we should try to move back to a debian version that worked. But that would probably imply using a debian version from the archive.

hammera commented:
@egli, I full agree now you why removed from 3.27 milestone list this change, september 12 is very near, and need a thorough test to not risk the cross-compiled builds stability.
So, this is the correct and safe solution now. After 3.28 development cicle opens, we investigate this issue detailed with you wrote.
My experiences are exe files is fine, only test files not ran correct with Wine (letterdeftest_harness.yaml file).
I will looks my latest branch Docker file cross-compiled exe files with a real Windows system, now my friend brought me a windows machine for repair. :-):-)

Attila

hammera commented:
Ok, doed the wanted manual tests after 64 bit version compiled right too.
32 bit build version results, with Wine tests need exporting LOUIS_TABLEPATH environment value my machine for example with z:\\home\\hammera\\liblouis\\share\\liblouis\\tables value:
* wine bin/lou_translate.exe --version right shows the copyright output.
* wine bin/lou_translate.exe hu-hu-g1.ctb command right translates newest hungarian exceptions.
* wine bin/lou_checkyaml.exe hu-hu-g1_dictionary_special_harness.yaml large test file testing ran successful (of course me need copying the test file from the source repository).

64 bit version test results:
* wine bin/lou_translate.exe --version right shows the copyright output.
* wine bin/lou_translate.exe hu-hu-g1.ctb command right translates newest hungarian exceptions.
* wine bin/lou_checkyaml.exe hu-hu-g1_dictionary_special_harness.yaml large test file testing ran successful (of course me need copying the test file from the source repository).

Attila

hammera commented:
Sorry boys, but I have one more question:
Should I remove the sha256sum values from the Docker files in the update_docker_files development branch? The sha256sum value has changed again (from f2150eb to b91baba), there is also an automatic pull request opened by @dependabot in this regard. The bot opened with pr #1441.
Do we want to constantly have fun merging these little automatic opened very trivial changes? :-):-)
Isn't it better to simply use debian:latest, debian:bookworm or debian:bullseye for the pull lines? This type of code always pulls the latest sha256sum checksum image from the Docker hub, depends we what code name to use.
debian:latest equals with always the latest stable distribution, debian:bookworm now the actual stable version too, debian:bullseye is the old stable code name distribution, after releasing next or next+1 stable version, will be unavailable this version related packages the package repository. This situation if this is happening, and bulleseye get end of support time (eol time), not possible downloading any packages from apt repository the bullseye code name related, because any downloads future will be result 404 status code if bullseye is end of life. We few month older time happened similar issue with debian:jessie code name based liblouis:liblouis image, when we want do a test Liblouis UTDML build.
Sorry if this part not good my english, but perhaps understand what want I say. Bookworm have more time support with Bullseye old stable code name.

For Liblouis UTDML Docker files, e.g. we don't mess with sha256sum checksums, we just pull the latest disk image from the Docker hub.

egli commented:
The sha sums are for security reasons, so that we define what version we want. Otherwise upstream just upgrades and we are automatically upgrading whether we want or not.

hammera commented:
@egli, ok, agree. :-):-)
This is a suggestion only with I wrote.

What sha256sum value I use? f2150eb or b91baba? Or future the bot will be handle this?
Debian:latest code name is good all docker files, or would like different code name?


If sha256sum is need, future we need fixing this with Liblouis UTDML too the Dockerfile.win32 and Dockerfile.win64 files, because this project we not use sha256sum values. Need define this situation an automatic both mechanishm with always updates the sha256sum values this files if happening sha256sum changes, I don't no in Liblouis this mechanishm are how configured previous.

Attila

hammera commented:
Boys, I see an another PR the milestone list with want's pinning the Docker files with Bullseye based images (the Bullseye based image related pull request already getted the approve flag).
So wat we doing this PR? Simple I close this pull request (to have possibility doing the Bullseye based pull request)?

Attila

egli commented on Dockerfile:
```diff
@@ -3,7 +3,7 @@ FROM debian:latest@sha256:f2150eba68619015058b26d50e47f9fba81213d1cb81633be7928c
 
 # Fetch build dependencies
-RUN apt-get update && apt-get install -y \
+RUN apt-get update && apt-get -y dist-upgrade && apt-get install -y \
```
We should not do a `dist-upgrade`. We specify the Debian version using the `FROM` keyword

egli commented on Dockerfile:
```diff
@@ -14,7 +14,7 @@ RUN apt-get update && apt-get install -y \
     python3-distutils \
     texinfo \
-   && rm -rf /var/lib/apt/lists/*
+   && apt-get clean && rm -rf /var/lib/apt/lists/*
```
As far as I know the `rm -rf /var/lib/apt/lists/*` should be enough. It seems to be common practice

egli commented on Dockerfile.win32:
```diff
@@ -62,21 +62,22 @@ RUN ./autogen.sh && \
-# just run any old self-contained yaml test that doesn't need any env setup
-ADD tests/yaml/letterDefTest_harness.yaml .
-RUN wine bin/lou_checkyaml.exe letterDefTest_harness.yaml
+# run spaces.yaml test to we ensure cross-compile build works right with 32 bit Windows versions
```
@hammera I think you explained this somewhere else but I don't really understand why the tests should work when the spaces test instead of the letterDefTest_harness test. If the spaces test fails we should find out why instead of just using a different test.

egli commented on Dockerfile.win64:
```diff
@@ -42,7 +42,7 @@ RUN ./autogen.sh && \
         CPPFLAGS="-I${PREFIX}/libyaml/include/" LDFLAGS="-L${PREFIX}/libyaml/lib/" && \
     make LDFLAGS="-L${PREFIX}/libyaml/lib/ -avoid-version -Xcompiler -static-libgcc" && \
-    make check WINE=wine64 || cat tests/test-suite.log && \
+    make check WINE=/usr/lib/wine/wine64 || cat tests/test-suite.log && \
```
To be honest I'm not sure if that is the correct way to invoke wine64

egli reviewed (changes requested):
Hi @hammera I'm not sure what to do with this PR. I listed some sticky points in the comments.

- the dist-upgrade and the hash numbers can be dropped. I guess they hve been superceded by the other PR
- changing the letterDefTest_harness with spaces test seems random. Why would that help?
- invoking wine64 via `/usr/lib/wine/` feels weird. Is that really how you have to do it?

hammera commented:
Ok Chris, I begin understanding what you would like.
The tests/yaml/letterdefharness.yaml test file replacement is happened because this test produces many mistific error messages the test logs during cross-compilation before the yaml test system shows succes result, and I want ensure tests ran correctly an another test (for example with spaces.yaml).
Letterdeftest_harness.yaml shows following errors for example my Linux box with latest master branch version, but I don't no why:
```
source unknown: error: Character can not be (indirectly) based on itself.
source unknown: error: Character can not be (indirectly) based on itself.
Translation failed.
SUCCESS (8 tests, 0 failures)
```
The dist-upgrade related part, and another apt-get parts related you are right, for example apt-get dist-upgrade really not need if we pinning a specific bullseye image.
Apropo, bullseye or latest images:
Have a missing change in Dockerfile. Dockerfile uses debian:latest based image (in bookworm apt repository), but Dockerfile.win32, Dockerfile.win64 uses bullseye based images.
Commit the proper bullseye based image into the Dockerfile from Dockerfile.win32 file?

With wine64 execute related:
I uses the /usr/lib/wine/wine64 I think, because wine64 not in the path the 64 bit builded Docker image (perhaps not have a simlink in /usr/bin, I don't no why). So when I tryed execute simple wine64 command, bash not found the wine64 binary (in /usr/bin or /bin not exists with wine64 binary and not have any simlink this binary).
The 32 wine build is not affected.

If everything is clear now, I actualizing my branch your requested changes (and fix the Dockerfile pull line back with Bullseye based image, similar with you already doed and merged with Dockerfile.win32 and Dockerfile.win64 command).

Attila

hammera commented:
Ok, now I doed olni a rebase to actualize the feature branch.
The third commit perhaps we have possibility to drop out (with optimizations related changes, Chris absolute right).
So apt-get dist-upgrade, apt-get clean parts related change are real unneed.
The last, normal Dockerfile related change is need (I synced back with Bullseye based image from Dockerfile.win32 file, normal, copy, paste, commit task happened only).
Only have one thing remaining, what we doing the test file replacement related part, and what we doing with 64 bit wine build related the wine execute command part.

Attila

egli commented:
I built the Docker images with the Dockerfiles that are based on bullseye and the tests passed. I get some weird error messages but I think they are related to the wine installation, not something we really need to worry about.

The method of building with the bullseye docker image will work [until June 30th, 2026](https://wiki.debian.org/LTS) so I have very little incentive to change anything about these Dockerfiles right now.

Maybe if we had a PR that would migrate the Dockerfiles to Debian latest I'd be willing to merge.

Also after having talked to Bert I no longer have a problem with 

1. taking wine64 from /usr/lib/wine/wine64
2. using the spaces test instead of letterDefTest

hammera commented:
Thanks Chris the log quote, I see now the log file the interesting part.
So perhaps in Bullseye not need use the /usr/lib/wine/wine64 command, simple wine64 command execute works right and the yaml file final ran succesfull.

I don't understand following part what you would like:
„Maybe if we had a PR that would migrate the Dockerfiles to Debian latest I'd be willing to merge.”
When you would like migrate Docker files to debian:latest? Now we goed back with Bullseye based images. :-):-)
Original, the update_docker_files branch I used debian:latest images (this images uses Debian Bookworm release), but because you final stepped back with debian:bullseye based Docker images, I sync back yesterday all Docker files with this image with you using. :-):-)
So, I think possible misunderstand the purpose. In 2023 june 5 temporary I think all Docker files migrated with debian:latest image (perhaps after Liblouis 3.26.0 release).
Look with following commit:
„commit 56f26d8b39b0d9370e84a5f176c3c5af88d4182b
Author: Christian Egli <christian.egli@sbs.ch>
Date:   Mon Jun 5 17:19:06 2023 +0200

    Update the Dockerfiles to use latest Debian”
In I think august 11 you doed an another fix with Dockerfile related, actualized some packages.
Only the cross-compilation not worked with original Dockerfile.win64 if you not used with /usr/lib/wine/wine64 full command if you used with debian:latest image (because in /usr/bin not exists in debian:latest image the wine64 binary or not have a simlink).
This trick are not need in Bullseye. :-):-)
Until june 30, 2026 Bullseye is supported, you are absolute right.

So:
What we do now this PR? Simple Close because obsoleted?
If we want staying in Bullseye until end of Bullseye support, this is right because not have risks, (have estimated 2,5 year support), only this situation the Dockerfile need update the image pull line, other master branch have Docker files is good (Dockerfile.win32 and Dockerfile.win64), and cross-compilation happens right with 32 bit and 64 bit Windows builds.

Before june 30, 2026 we will migrate to debian:bookworm image, or debian:latest image (depending the choice with Bert and you how would like and final what image wan'ts use).

I will begin prepare this change in 2026. january (or perhaps 2025. november or december), to we have enough time refactoring the migration related tasks.

What do you think?

Attila

hammera commented:
Boys, I see this pull request are scheduled with 3.29 release.
Because now Dockerfile.win32 and Dockerfile.win64 are use the Bullseye based image, but Dockerfile is not (the Linux related build creator image uses debian:latest image), I need me doing a simple Pull request with fixing only the Dockerfile pull line to Bullseye based image?
Cross-compilation is not affected, so you already easy to do Liblouis 32 bit and 64 bit Windows build again (Chris fixed this), but if you want's Linux based Liblouis docker image, perhaps better to fix the Dockerfile pull line equals with Bullseye based image.
The my branch doed simple last commit fixing this image mismatch (other commits is interesting only if we future want doing Liblouis builds with Debian:bookworm latest stable release based image after 2026. june 30).
Better to future wee keep codenames related pull lines (debian:bookworm, debian:trixie if changing the stable release), because if future Debian releasing again a new stable release, we not have risks, because if we not using debian:latest pull lines, Docker not will be pulling automatically  the latest stable Debian base image and not will be have unwanted regression or package incompatibilities with Docker files related during cross-compilation.
So, if we always using code lines the pull line, minimize the risks.

Attila

egli commented:
Hi Attila

I think that we will stay with bullseye for cross compilation for as long as possible (`Dockerfile.win32` and `Dockerfile.win64`). 

The normal `Dockerfile` is just used to run liblouis if do not want to install it. This one is not used for cross compilation and has no wine and all the cross compilation tools in it. So that can as far as I'm concerned easily be based on Debian latest.

Now if someone provides me with a PR that upgrades the cross compilation Docker images so that

1. the cross compilation, `make test` and the test with `wine` still works
2. there is no weird interaction with wine (I'm still not sure if calling `wine64` from `/usr/lib/` is really kosher)

then I'd be happy to move to a newer Debian than bullseye.

But for now it works and I'd rather work on other issues.

With that in mind I think and if you agree we can close this issue.

hammera commented:
Ok Chris, full agree with you wrote.
Feel free to close this issue and concern other issues.
Before the end of the support period for the Bullseye edition, I will deal with how to solve the cross-compilation in the case of Windows versions in such a way that the solution fully meets the professional criteria you set.
/usr/lib/wine/wine64 was a forced solution, I also know that this approach is not entirely correct, but the tests in the Docker container cannot be performed in the case of cross-compilation on the Bookworm branch without I not use the full path of wine64 binary path. :-):-)
Hopefully this will be resolved by the time the Bullseye release expires in the current Debian stable release.
Perhaps if we future use a soft simlink from /usr/lib/wine/wine64 to /usr/bin/wine64 with ln -s command, this is a more elegant solution if not have better fix with distribution stable suite or wine package level.
So:
Because this issue now is obsoleted, anybody feel free to close this issue the proper flag (me simple have possibility only to clicking close button the pull request dialogue), but before close the issue please remove this issue from the 3.29.0 milestone cicle to this closed issue not present the closed issue list.

Attila
